django/db/models/query.py

```diff
@@ -1120,36 +1118,27 @@ def _merge_known_related_objects(self, other):
         for field, objects in other._known_related_objects.items():
             self._known_related_objects.setdefault(field, {}).update(objects)
 
-    def _prepare(self, field):
-        if self._fields is not None:
-            # values() queryset can only be used as nested queries
-            # if they are set up to select only a single field.
-            if len(self._fields or self.model._meta.concrete_fields) > 1:
-                raise TypeError('Cannot use multi-field values as a filter value.')
-        elif self.model != field.model:
-            # If the query is used as a subquery for a ForeignKey with non-pk
-            # target field, make sure to select the target field in the subquery.
-            foreign_fields = getattr(field, 'foreign_related_fields', ())
-            if len(foreign_fields) == 1 and not foreign_fields[0].primary_key:
-                return self.values(foreign_fields[0].name)
-        return self
-
-    def _as_sql(self, connection):
-        """
-        Returns the internal query's SQL and parameters (as a tuple).
-        """
-        if self._fields is not None:
+    def _prepare_as_filter_value(self):
+        if self._fields is None:
+            obj = self.values('pk')
+        else:
             # values() queryset can only be used as nested queries
             # if they are set up to select only a single field.
-            if len(self._fields or self.model._meta.concrete_fields) > 1:
+            if len(self._fields) > 1:
                 raise TypeError('Cannot use multi-field values as a filter value.')
-            clone = self._clone()
-        else:
-            clone = self.values('pk')
-
-        if clone._db is None or connection == connections[clone._db]:
-            return clone.query.get_compiler(connection=connection).as_nested_sql()
-        raise ValueError("Can't do subqueries with queries on different DBs.")
+            obj = self._clone()
+        query = obj.query
+        query._db = obj._db
+        query.subquery = True
+        if self._fields is None:
+            query._forced_pk = True
```

Comment:
Could this assignment be moved to the previous `if self._fields is None` check at the beginning of the method? Seems strange to have this down here, even though this is the place you're operating on the query object. Still, a `obj.query._forced_pk = True` would probably help reading.